### PR TITLE
Add geocoding to display city in reports

### LINKED ISSUE #77: Use 'geocoding' package for displaying geo information
Displaying geo information was removed on the pull request: https://github.com/Mosquito-Alert/Mosquito-Alert-Mobile-App/pull/34

Try to keep displaying that information by using the `placemarkFromCoordinates` from the `geocoding` library. (https://pub.dev/packages/geocoding) package instead.

Things that should display geo information again (that was removed on the pull request):

---

1. Show geo information on report list page.

https://github.com/Mosquito-Alert/Mosquito-Alert-Mobile-App/blob/6c5993a230a86f958c8dca8bcfef2994a6b93ebe/lib/pages/my_reports_pages/my_reports_page.dart#L558

Should append: `"{CITY} ({SUB_ADMIN_AREA})"`

---

2. model `Report.displayCity` should be filled with the CITY name asynchronously.

It's used here: https://github.com/Mosquito-Alert/Mosquito-Alert-Mobile-App/blob/6c5993a230a86f958c8dca8bcfef2994a6b93ebe/lib/pages/my_reports_pages/components/reports_list_widget.dart#L68

### PR DESCRIPTION
**Pull request by DigitalSeneca**
fixes #77 

| 1 | 2 |
| ------------- | ------------- |
| image  | image  |

### CODE REVIEW

File: lib/pages/forms_pages/components/biting_location_form.dart [hunk 1]
```diff
@@ -47,7 +47,8 @@ class _BitingLocationFormState extends State<BitingLocationForm> {
                   Utils.report!.selected_location_lon!)));
           currentLocation = Position(
               latitude: Utils.report!.selected_location_lat!,
-              longitude: Utils.report!.selected_location_lon!);
+              longitude: Utils.report!.selected_location_lon!,
+              accuracy: -1, altitude: -1, altitudeAccuracy: -1, speed: -1, speedAccuracy: -1, heading: -1, headingAccuracy: -1, timestamp: null);
```
Comment (DigitalSeneca) [inline lib/pages/forms_pages/components/biting_location_form.dart:51]: I got the error that all parameters are required and I could not compile otherwise. I am not sure why this was not happening before, probably some version/dependencies stuff
```
The named parameter 'speed' is required, but there's no corresponding argument.
Try adding the required argument.dart[missing_required_argument](https://dart.dev/diagnostics/missing_required_argument)
```
image

File: lib/utils/Utils.dart [hunk 2]
```diff
@@ -104,7 +104,7 @@ class Utils {
       }
     }
 
-    if (session!.id != null && language != null) {
```
Comment (DigitalSeneca) [inline lib/utils/Utils.dart:107]: dart linter: "language can't be null, remove useless comparison"

File: lib/api/api.dart [hunk 3]
```diff
@@ -779,4 +793,10 @@ class ApiSingleton {
       return false;
     }
   }
+  
+  Future<String> getCityNameFromCoords(double lat, double lon) async {
+    var locale = '${await UserManager.getLanguage()}_${await UserManager.getLanguageCountry()}';
```
Comment (DigitalSeneca) [inline lib/api/api.dart:798]: Locale in geocoding is optional, if not provided it uses the one that the phone has selected
If we add it, it has to be in the formal language_country (ie es_ES, de_DE, en_US...)

File: lib/api/api.dart [hunk 4]
```diff
@@ -779,4 +793,10 @@ class ApiSingleton {
...
+    var placemarks = await placemarkFromCoordinates(lat, lon, localeIdentifier: locale);
+    return placemarks[0].locality ?? '';
```
Comment (DigitalSeneca) [inline lib/api/api.dart:800]: If more than one placemark is returned, all of them will be very close to each other, so I'd just take the first one

File: lib/api/api.dart [hunk 5]
```diff
@@ -331,6 +332,9 @@ class ApiSingleton {
       if (report.selected_location_lon != null) {
         body.addAll({'selected_location_lon': report.selected_location_lon});
       }
+      if (report.displayCity != null) {
+        body.addAll({'display_city': report.displayCity});
+      }
```
Comment (epou) [inline lib/api/api.dart:337]: The API does not accept this parameter... Need to be removed. In the future we will make the backend fetch the city name and expose it through the API.

File: lib/api/api.dart [hunk 6]
```diff
@@ -449,6 +453,16 @@ class ApiSingleton {
...
+          double lat, lon;
+          if (item['location_choice'] == 'current') {
+              lat = item['current_location_lat'] ?? Utils.defaultLocation.latitude;
+              lon = item['current_location_lon'] ?? Utils.defaultLocation.longitude;
+          } else {
+              lat = item['selected_location_lat'] ?? Utils.defaultLocation.latitude;
+              lon = item['selected_location_lon'] ?? Utils.defaultLocation.longitude;
+          }
+          var cityName = await getCityNameFromCoords(lat, lon);
+          item['display_city'] = cityName;
```
Comment (epou) [inline lib/api/api.dart:465]: I suggest relocating this code within the Report constructor as it seems more fitting there. This logic appears unrelated to the communication with the API.

File: lib/api/api.dart [hunk 7]
```diff
@@ -779,4 +793,10 @@ class ApiSingleton {
       return false;
     }
   }
+  
+  Future<String> getCityNameFromCoords(double lat, double lon) async {
```
Comment (epou) [inline lib/api/api.dart:797]: I think it'd be better to place this function in the Utils folder since it's note related to how to communicate with the API.

File: lib/api/api.dart (hunk 3 above)
Comment (epou) [reply lib/api/api.dart:798]: Would you mind creating a function in the UserManger called getUserLocale (or something like this) that returns this value? It could be worth reusing that function in the future for other cases instead of contacting the result of getLanguage and getLanguageCountry always

Comment (epou) [reply lib/api/api.dart:798]: It would be nice to check if it works in basque... I was wondering if eu_ES is the correct locale code. If it's prone to error, only using language is also accepted in the function `placemarkFromCoordinates`

File: lib/api/api.dart (hunk 4 above)
Comment (epou) [reply lib/api/api.dart:800]: Need to check if placemarks is empty before getting the first position:
```dart
if (placemarks.isNotEmpty) {
    // code here
 }
 ```

File: lib/pages/forms_pages/components/biting_location_form.dart (hunk 1 above)
Comment (epou) [reply lib/pages/forms_pages/components/biting_location_form.dart:51]: Ah, that's interesting. It appears that the `Position` class from the `geolocator` package wasn't designed to be instantiated directly. It serves as the type of object returned by methods like `getCurrentPosition` or `getLastKnownPosition`. Considering this, it might be more appropriate to replace them all and utilize the `LatLng` class from `google_maps_flutter.dart` instead, which seems to be used elsewhere in the project. However, making this change is significant and requires thorough testing on the device to ensure that everything continues to function as intended.

File: lib/api/api.dart (hunk 4 above)
Comment (DigitalSeneca) [reply lib/api/api.dart:800]: "locality" can be null, for example if it's in the middle of the ocean (0, 0). In this case I assume in the app it will just not show anything. I can modify the method signature to be able to return null values:
`Future<String?> getCityNameFromCoords(double lat, double lon) async {`

Another option is to do something like this to ensure that always either the locality or an empty string is returned
```
Future<String> getCityNameFromCoords(double lat, double lon) async {
    var locale = '${await UserManager.getLanguage()}_${await UserManager.getLanguageCountry()}';
    var placemarks = await placemarkFromCoordinates(lat, lon, localeIdentifier: locale);
    if (placemarks.isEmpty) { return ''; }
    return placemarks.first.locality ?? '';
}
```

File: lib/api/api.dart (hunk 6 above)
Comment (DigitalSeneca) [reply lib/api/api.dart:465]: I agree, I tried first to do it first in the Report.fromJson constructor but got:
```
The modifier 'async' can't be applied to the body of a constructor.
Try removing the modifier.dart[invalid_modifier_on_constructor](https://dart.dev/diagnostics/invalid_modifier_on_constructor)
```

File: lib/api/api.dart (hunk 4 above)
Comment (epou) [reply lib/api/api.dart]: Yeah, returning always String looks good to me

File: lib/api/api.dart (hunk 3 above)
Comment (DigitalSeneca) [reply lib/api/api.dart]: > Would you mind creating a function in the UserManger called getUserLocale (or something like this) that returns this value?

done

> It would be nice to check if it works in basque... I was wondering if eu_ES is the correct locale code. If it's prone to error, only using language is also accepted in the function placemarkFromCoordinates

Seems like it's not working as expected. It's returning the name in the countries name instead of in selected locale (Mountain View in US; München in Germany, and Bilbo in Basque), when I have basque as my language in the app
image

Comment (DigitalSeneca) [reply lib/api/api.dart]: > Or try using official method: https://api.flutter.dev/flutter/intl/Intl/getCurrentLocale.html

```
static String getCurrentLocale() {
  return defaultLocale ??= systemLocale;
}
```

But then with this method, if the user selects another language within the app, it will always still show the systemLocale

File: lib/api/api.dart (hunk 6 above)
Comment (epou) [reply lib/api/api.dart]: Have you tried using a separate async function inside the class and call it?

```dart
class Report {
 // properties here

  Report() {
   // current code here
    _initialize();
  }

  Future<void> _initialize() async {
   // code to get appropiate lat and lon here
    var cityName = await getCityNameFromCoords(lat, lon);
    displayCity = cityName;
  }
}
```

File: lib/api/api.dart (hunk 3 above)
Comment (epou) [reply lib/api/api.dart]: Ah yes, you're right! The only thing I'm afraid of is the `UserManger.getUserLocale` returning a combination of language + country that is not a valid locale and the code fails. Maybe it's safer to just use the language code as stated in the [docu](https://pub.dev/packages/geocoding) ('en' for English)

Comment (DigitalSeneca) [reply lib/api/api.dart]: > The localeIdentifier should be formatted using the syntax: [languageCode]_[countryCode]

`en` seems to be the only option where a country code is not needed. If we are afraid, maybe the safest option is to leave this option parameter empty, so it will use the system's language

Comment (epou) [reply lib/api/api.dart]: As far as this codes are always valid, it's gonna be ok
https://github.com/Mosquito-Alert/Mosquito-Alert-Mobile-App/blob/6c5993a230a86f958c8dca8bcfef2994a6b93ebe/lib/pages/settings_pages/settings_page.dart#L38-L62

File: lib/api/api.dart (hunk 6 above)
Comment (DigitalSeneca) [reply lib/api/api.dart]: I just tried it and it works, nice. I agree that's better than it being in the api.dart

File: lib/pages/forms_pages/components/biting_location_form.dart (hunk 1 above)
Comment (DigitalSeneca) [reply lib/pages/forms_pages/components/biting_location_form.dart:51]: yup, LatLng seems more appropriate when the other parameters are not used

> However, making this change is significant and requires thorough testing on the device to ensure that everything continues to function as intended.

Are we doing it in a future version then?

Comment (epou) [reply lib/pages/forms_pages/components/biting_location_form.dart:51]: Could you try replacing all "Position" instances for LatLng? Maybe it's just that simple. Test it in your emulator just to be sure, but we can test it later once the draft is released

Comment (DigitalSeneca) [reply lib/pages/forms_pages/components/biting_location_form.dart]: I just tried it and seems to be working on my emulator. You can see it in the latest commit "Replace Position for LatLng"

**Review by epou — APPROVED** [review]
